diff --git a/dlls/winhttp/request.c b/dlls/winhttp/request.c
index 60ba98da574..48c02528d7d 100644
--- a/dlls/winhttp/request.c
+++ b/dlls/winhttp/request.c
@@ -615,7 +615,8 @@ BOOL WINAPI WinHttpAddRequestHeaders( HINTERNET hrequest, const WCHAR *headers,
     return !ret;
 }
 
-#define QUERY_MODIFIER_MASK (WINHTTP_QUERY_FLAG_REQUEST_HEADERS | WINHTTP_QUERY_FLAG_SYSTEMTIME | WINHTTP_QUERY_FLAG_NUMBER)
+#define QUERY_MODIFIER_MASK (WINHTTP_QUERY_FLAG_REQUEST_HEADERS | WINHTTP_QUERY_FLAG_SYSTEMTIME | WINHTTP_QUERY_FLAG_NUMBER \
+                             | WINHTTP_QUERY_FLAG_NUMBER64)
 
 static DWORD query_headers( struct request *request, DWORD level, const WCHAR *name, void *buffer, DWORD *buflen,
                             DWORD *index )
@@ -747,6 +748,23 @@ static DWORD query_headers( struct request *request, DWORD level, const WCHAR *n
         }
         *buflen = sizeof(DWORD);
     }
+    else if (level & WINHTTP_QUERY_FLAG_NUMBER64)
+    {
+        if (attr != WINHTTP_QUERY_CONTENT_LENGTH)
+        {
+            WARN( "WINHTTP_QUERY_FLAG_NUMBER64 with attr %lu.\n", attr );
+            return ERROR_WINHTTP_INVALID_QUERY_REQUEST;
+        }
+        if (!buffer || sizeof(ULONG64) > *buflen) ret = ERROR_INSUFFICIENT_BUFFER;
+        else
+        {
+            ULONG64 *number = buffer;
+            *number = wcstoull( header->value, NULL, 10 );
+            TRACE( "returning number: %I64u\n", *number );
+            ret = ERROR_SUCCESS;
+        }
+        *buflen = sizeof(ULONG64);
+    }
     else if (level & WINHTTP_QUERY_FLAG_SYSTEMTIME)
     {
         SYSTEMTIME *st = buffer;
